File: leetcode/tree/2277_Closest_Node_to_Path_in_Tree/codes/dfs_bfs.py
from collections import deque, defaultdict
from typing import List

class Solution:
    def closestNode(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        def bfs_dist(src):
            dist = [-1] * n
            q = deque([src])
            dist[src] = 0
            while q:
                node = q.popleft()
                for nei in graph[node]:
                    if dist[nei] == -1:
                        dist[nei] = dist[node] + 1
                        q.append(nei)
            return dist

        def dfs_path(u, target, parent, path):
            path.append(u)
            if u == target:
                return True
            for v in graph[u]:
                if v != parent:
                    if dfs_path(v, target, u, path):
                        return True
            path.pop()
            return False

        res = []
        for start, end, node in query:
            path = []
            dfs_path(start, end, -1, path)
            dist = bfs_dist(node)
            closest = min(path, key=lambda x: dist[x])
            res.append(closest)
        return res

File: leetcode/tree/2277_Closest_Node_to_Path_in_Tree/codes/dfs_bfs.cpp
#include <vector>
#include <queue>
#include <climits>
using namespace std;

class Solution {
public:
    vector<vector<int>> graph;

    vector<int> closestNode(int n, vector<vector<int>>& edges, vector<vector<int>>& query) {
        graph = vector<vector<int>>(n);
        for (auto& e : edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }

        vector<int> res;
        for (auto& q : query) {
            int start = q[0], end = q[1], node = q[2];
            vector<int> path;
            dfsPath(start, end, -1, path);

            vector<int> dist = bfsDist(n, node);
            int minNode = path[0];
            for (int p : path) {
                if (dist[p] < dist[minNode]) {
                    minNode = p;
                }
            }
            res.push_back(minNode);
        }

        return res;
    }

    void dfsPath(int u, int target, int parent, vector<int>& path) {
        path.push_back(u);
        if (u == target) return;
        for (int v : graph[u]) {
            if (v != parent) {
                int prevSize = path.size();
                dfsPath(v, target, u, path);
                if (path.back() == target) return;
                path.resize(prevSize);
            }
        }
    }

    vector<int> bfsDist(int n, int src) {
        vector<int> dist(n, -1);
        queue<int> q;
        q.push(src);
        dist[src] = 0;
        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int v : graph[u]) {
                if (dist[v] == -1) {
                    dist[v] = dist[u] + 1;
                    q.push(v);
                }
            }
        }
        return dist;
    }
};
